### DISCUSSION TITLE
noMasked changes beta values

### DISCUSSION THREAD

**Discussion by NazaninMir** [zwdzwd/sesame#167] [Q&A]
Hi all, 
I'm currently working on an EPICv2 array using sesame .. during the QC steps I realized that the beta-values change after applying the noMasked function on the SigDF. Anyone else experienced the same? or can someone explain this ?

**Comment by zwdzwd** [comment]
If you apply getBetas directly, like getBetas(noMasked(sdf)) and getBetas(sdf), there shouldn't be a difference. My guess is you are using openSesame which includes additional preprocessing some of which may use the masked probes.

**Comment by NazaninMir** [comment]
Thank you for your reply. 
Actually I didn't used openSesame. 
Thats my command: 
sdfMask <- lapply(sdf, function (x) noMasked(x))

**Comment by zwdzwd** [comment]
Is sdf just one SigDF? under the hood noMasked is simply
```
sdf[!sdf$mask,,drop=FALSE]
```
so you just need
```
noMasked(sdf)
```

**Comment by NazaninMir** [comment]
No I have over 50  SigDF in sdf.
List of 56
 $ 207222480102_R01C01:Classes ‘SigDF’ and 'data.frame':	937690 obs. of  7 variables:
....
I use lapply, cause I could not figure out wether other processes are running when using openSesame

**Reply by zwdzwd** [reply]
OK. then how did you check beta values and what else did you do before that?

**Comment by NazaninMir** [comment]
Here are the steps I applied. I did all QC step by step,  to be able to reconstruct:
idatlist <- lapply(idats, function (x) readIDATpair(x, manifest = addr))
test.betas <- do.call(cbind, lapply(idatlist,function (x) getBetas(x, mask=TRUE)))

#inferInfiniumIChannel
idatlist_C <- lapply(idatlist, inferInfiniumIChannel)
test.betasC <- do.call(cbind, lapply(idatlist_C,function (x) getBetas(x, mask=TRUE)))

 
# Run dye bias correction
idatlist_CD <- lapply(idatlist_C, dyeBiasNL)
test.betasCD <- do.call(cbind, lapply(idatlist_CD,function (x) getBetas(x, mask=TRUE)))

#Detection Pvalue
idatlist_CDPo <- lapply(idatlist_CD, function (x) pOOBAH(x, pval.threshold = 0.01))
test.betasCDPo <- do.call(cbind, lapply(idatlist_CDPo,function (x) getBetas(x, mask=TRUE)))

# -> until here everything seems fine. Beta values are as they should be
#####################################################################

#-> after this step some beta values changes.
 
idatlist_CDPo_ExcMask <- lapply(idatlist_CDPo,function (x) noMasked(x)) 
test.betasCDPoMask <- do.call(cbind, lapply(idatlist_CDPo_ExcMask,function (x) getBetas(x, mask=TRUE)))

**Reply by zwdzwd** [reply]
thanks. looking at the code https://github.com/zwdzwd/sesame/blob/724d4c5e7f70dc903d1a4145c2758e5c80f412c5/R/sesame.R#L189
The only change expected is the filtering of masked probes and hence shorter dimension but there should be no change to the value that survive masking. or am I missing something?

**Reply by NazaninMir** [reply]
There should not be a change in the values .. and that is also the case in all the steps before .... 

test.betasCDPo[c("cgXXXX_BC11", "cgXXXY_TC21", "cgXXXZ_BC21"),][,1:3]
                207222480102_R01C01 207222480102_R02C01 207222480102_R03C01
cg00054210_BC11           0.4165562           0.4956148           0.5059058
cg00327322_TC21           0.3515975           0.3787637           0.4452577
cg01203153_BC21           0.1617281           0.1447128           0.1514285

But after running NoMasked some values change ... Not all but some . I think there is some kind of re- arrangement that messes up the data.

test.betasCDPoMask[c("cgXXXX_BC11", "cgXXXY_TC21", "cgXXXZ_BC21"),][,1:3]
                207222480102_R01C01 207222480102_R02C01 207222480102_R03C01
cg00054210_BC11           0.8903083           0.6598001           0.8574567
cg00327322_TC21           0.9022134           0.7865371           0.6779631
cg01203153_BC21           0.8391186           0.4313635           0.1078993

What I also realized is that, if a CpG-site has NAs for all samples It is not always dropped.. but rather somehow imputed!? 


test.betasCDPo[7:8,]
    207222480102_R01C01 207222480102_R02C01 207222480102_R03C01 207222480102_R04C01 207222480102_R05C01 
cg00000289_TC21                  NA                  NA                  NA                  NA                  NA                  

test.betasCDPoMask[7:8,]
                207222480102_R01C01 207222480102_R02C01 207222480102_R03C01 207222480102_R04C01 207222480102_R05C01 207222480102_R06C01
cg00000289_TC21           0.8637285            0.832685           0.7981342           0.8453972           0.8275779           0.8382574

**Reply by zwdzwd** [reply]
I think it might have to do with the fact after getBetas(noMasked()), different samples are no longer of the same length. Could it be that cbind is causing the mislabeling of probe IDs?

**Reply by NazaninMir** [reply]
Im sorry, somehow my answer was not sent. 
So here again: 

I tried also the NoMasked with openSesame and I get also changed values. 

I tried getBeta() on one sample with the and without the "NoMasked" function before and the values are changed. 
Correctly the one ( cg00000289_TC21 ) with an NA is excluded, but the CpG-values are changed.


OneSample <- getBetas(idatlist_CDPo[[1]], mask=TRUE)
-->
cg00000158_BC21 cg00000165_TC21 cg00000221_BC21 cg00000236_TC21 cg00000289_TC21 cg00000292_BC21 
0.9033180       0.2462405       0.8533618       0.7455211              NA       0.8637285 
cg00000321_BC21 
      0.2538314 

datlist_CDPo_ExcMask_open <- openSesame(idatlist_CDPo, func = noMasked,manifest=m2)
OneSample_ExcMask_open <- getBetas(idatlist_CDPo_ExcMask_open[[1]], mask=TRUE)

-->
 cg00000158_BC21 cg00000165_TC21 cg00000221_BC21 cg00000236_TC21 cg00000292_BC21 cg00000321_BC21 
0.9601634       0.1548068       0.9321163       0.8224250       0.9158312       0.2138720 
cg00000363_TC21 
      0.3206771

**Reply by NazaninMir** [reply]
Ok here I tried  NoMasked on one sample:
As you suggested it seems that using NoMasked with openSesame OR apply changes the structure of the list.

--> running both steps on one sample only dosen't change  the values .. but it also do not exclude the NA

onesample <- noMasked(idatlist_CDPo[[1]])
onesampleBetas <- getBetas(exonesample, mask=TRUE)

cg00000158_BC21 cg00000165_TC21 cg00000221_BC21 cg00000236_TC21 cg00000289_TC21 cg00000292_BC21 
0.9033180       0.2462405       0.8533618       0.7455211              NA       0.8637285 
cg00000321_BC21 
      0.2538314
